### ISSUE TITLE
IRI.create() builds erroneous parts when fragment starts with a number

### ISSUE THREAD

**Issue by andredd** [owlcs/owlapi#992]
IRI.create("prefix:1dontcare");

should return an IRI with namespace "prefix:" and fragment "1dontcare". Instead the namespace is "prefix:1" with fragment "dontcare"

**Comment by jamesaoverton** [comment]
I believe that OWLAPI is following the [QName](https://en.wikipedia.org/wiki/QName) specification here, which applies to XML element names and to identifiers in Turtle. QNames do not allow either of the two parts to start with a number.

**Comment by andredd** [comment]
Thanks for the reply james. In that case I think the API should produce some kind of error instead when confronted with an invalid input. The result shown above is at least quite counter-intuitive imho.

**Comment by ignazio1977** [comment]
`IRI::getFragment()` has Javadoc saying the string returned is the NCName fragment of the IRI, if one exists.

I don't see a reason for throwing an exception - the split between prefix and fragment is a convenience for human eyes and, to an extent, a consideration for memory footprint and diffs in serializations when changes are made to a file. But, throwing an exception would likely stop some tools from working with some ontologies. The IRI is always the full string, at RDF and OWL spec level.

The main reason for needing a split with certain rules at all is RDF/XML (cue rant: the overlap between awkward rules and RDF/XML related rules is very large), where properties must be represented as prefix+NCName tags, and so a property IRI and its split must satisfy certain rules.

There's another issue currently open about the fact that the NCName split causes some ontologies to have larger file serializations because many IRIs used in it have fragments starting with numbers; so, having the option would have a tangible positive effect. It's not easy to provide a solution that satisfies this need without causing issues, though, because IRI.create() has no knowledge of which kind of entity the IRI is for (classes, datatypes, named individuals could use either conventions regardless of usage and serialization language, but properties cannot; and changing conventions everywhere means existing files would see large diffs when changes are made with a newer OWLAPI version.

All of this could be accommodated by changing the serialization strategies, but this requires intervention on all renderers, so not a trivial change either.

Long story short, this "bug" has no easy fix, nor would the fix necessarily be welcomed by all users.

**Comment by andredd** [comment]
Thanks for the clarification Ignacio (which sounds absolutely reasonable to me). I'll work around the issue.

**Comment by Ayllonbe** [comment]
Hi @ignazio1977, sorry to restart that discussion, but I would like to ask if changes of the functions _isXMLNameStartCharacter()_ and _isXMLNameChar_ will be impacted in other functions than the IRI? I am saying that because If I modify them I can have the issue of erroneous parts fixed:

`public static boolean isXMLNameChar(int codePoint) {
    return isXMLNameStartCharacter(codePoint)
|| codePoint == '-' || codePoint == '.'
// || codePoint >= '0' && codePoint <= '9'             
|| codePoint == 0xB7
            || codePoint >= 0x0300 && codePoint <= 0x036F
            || codePoint >= 0x203F && codePoint <= 0x2040;
}`

`public static boolean isXMLNameStartCharacter(int codePoint) {
    return codePoint == ':' || codePoint >= 'A' && codePoint <= 'Z' || codePoint == '_'
            || codePoint >= '0' && codePoint <= '9'            || codePoint >= 'a' && codePoint <= 'z' || codePoint >= 0xC0 && codePoint <= 0xD6
            || codePoint >= 0xD8 && codePoint <= 0xF6 || codePoint >= 0xF8 && codePoint <= 0x2FF
            || codePoint >= 0x370 && codePoint <= 0x37D || codePoint >= 0x37F && codePoint <= 0x1FFF
            || codePoint >= 0x200C && codePoint <= 0x200D
            || codePoint >= 0x2070 && codePoint <= 0x218F
            || codePoint >= 0x2C00 && codePoint <= 0x2FEF
            || codePoint >= 0x3001 && codePoint <= 0xD7FF
            || codePoint >= 0xF900 && codePoint <= 0xFDCF
            || codePoint >= 0xFDF0 && codePoint <= 0xFFFD
            || codePoint >= 0x10000 && codePoint <= 0xEFFFF;
}`

Basically I changed  ´|| codePoint >= '0' && codePoint <= '9'` from a function to another. I tried with a test ontology where I include numerical values in different parts and seem to have the problem solved:

test#1: https://ontology.basf.net/ontology/2022/BASF_TEST/3TEST

namespace: https://ontology.basf.net/ontology/2022/BASF_TEST/
fragment: 3TEST

test#2: https://ontology.basf.net/ontology/2022/BASF_TEST/test#3TEST
namespace: https://ontology.basf.net/ontology/2022/BASF_TEST/test#
fragment: 3TEST

Please, I just do not know the impact on other functions but I can understand that it is not possible
